### ISSUE TITLE
GLIBC compatibility with RHEL/CentOS 7 and JRuby 9.4.0.0

### ISSUE THREAD

**Issue by li-afaris** [jruby/jruby#7579]
Hi. It appears the JNI bindings for JRuby 9.4.0.0 were built with GLIBC 2.27.  RHEL/CentOS 7 ships with GLIBC 2.17. Is it possible to support older builds of GLIBC with future builds of JRuby 9.4?  It would be great to use the features of MRI Ruby 3.1.  

Thanks

**Environment Information**
```
afaris:jruby-9.4.0.0 $ ./bin/jruby --version
jruby 9.4.0.0 (3.1.0) 2022-11-23 95c0ec159f OpenJDK 64-Bit Server VM 11.0.8+10-20200812 on 11.0.8+10-20200812 +jit [x86_64-linux]

afaris:jruby-9.4.0.0 $ cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)

afaris:jruby-9.4.0.0 $ rpm -q glibc
glibc-2.17-326.el7_9.i686
glibc-2.17-326.el7_9.x86_64
```
**Expected Behavior**
JRuby would not throw an exception when using JNI bindings. 

```
# executed on a fedora host with GLIBC 2.35
$  ./bin/jruby -Xnative.verbose -w -d --dev -e 'system "date"' 
Successfully loaded native POSIX impl.
Thu Jan 12 02:12:32 PM PST 2023
```

**Actual Behavior**
When running jruby 9.4.0.0 on centos 7 I get a glibc exception.

```
afaris:jruby-9.4.0.0 $  ./bin/jruby -Xnative.verbose -w -d --dev -e 'system "date"' 
Failed to load native POSIX impl; falling back on Java impl. Stacktrace follows.
java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
	at org.jruby.dist/jnr.ffi.provider.InvalidProvider$1.loadLibrary(InvalidProvider.java:49)
	at org.jruby.dist/jnr.ffi.LibraryLoader.load(LibraryLoader.java:420)
	at org.jruby.dist/jnr.posix.POSIXFactory$DefaultLibCProvider$SingletonHolder.<clinit>(POSIXFactory.java:307)
	at org.jruby.dist/jnr.posix.POSIXFactory$DefaultLibCProvider.getLibC(POSIXFactory.java:335)
	at org.jruby.dist/jnr.posix.BaseNativePOSIX.<init>(BaseNativePOSIX.java:41)
	at org.jruby.dist/jnr.posix.LinuxPOSIX.<init>(LinuxPOSIX.java:23)
	at org.jruby.dist/jnr.posix.POSIXFactory.loadLinuxPOSIX(POSIXFactory.java:149)
	at org.jruby.dist/jnr.posix.POSIXFactory.loadNativePOSIX(POSIXFactory.java:124)
	at org.jruby.dist/jnr.posix.POSIXFactory.loadPOSIX(POSIXFactory.java:93)
	at org.jruby.dist/jnr.posix.LazyPOSIX.loadPOSIX(LazyPOSIX.java:40)
	at org.jruby.dist/jnr.posix.LazyPOSIX.posix(LazyPOSIX.java:34)
	at org.jruby.dist/jnr.posix.LazyPOSIX.isNative(LazyPOSIX.java:404)
	at org.jruby.dist/org.jruby.util.io.FilenoUtil.<init>(FilenoUtil.java:43)
	at org.jruby.dist/org.jruby.Ruby.<init>(Ruby.java:298)
	at org.jruby.dist/org.jruby.Ruby.newInstance(Ruby.java:730)
	at org.jruby.dist/org.jruby.Main.internalRun(Main.java:266)
	at org.jruby.dist/org.jruby.Main.run(Main.java:227)
	at org.jruby.dist/org.jruby.Main.main(Main.java:199)
Caused by: java.lang.UnsatisfiedLinkError: could not get native definition for type `POINTER`, original error message follows: java.lang.UnsatisfiedLinkError: Unable to execute or load jffi binary stub from `/tmp`. Set `TMPDIR` or Java property `java.io.tmpdir` to a read/write path that is not mounted "noexec".
/home/afaris/jruby/jruby-9.4.0.0/jffi10117070039016966684.so: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by /home/afaris/jruby/jruby-9.4.0.0/jffi10117070039016966684.so)
	at org.jruby.dist/com.kenai.jffi.internal.StubLoader.tempLoadError(StubLoader.java:555)
	at org.jruby.dist/com.kenai.jffi.internal.StubLoader.loadFromJar(StubLoader.java:454)
	at org.jruby.dist/com.kenai.jffi.internal.StubLoader.load(StubLoader.java:330)
...
```

This is confirmed by the ldd command on centos 7. 
```
afaris:jruby-9.4.0.0 $ ldd lib/jni/x86_64-Linux/libjffi-1.2.so
ldd: warning: you do not have execution permission for `lib/jni/x86_64-Linux/libjffi-1.2.so'
lib/jni/x86_64-Linux/libjffi-1.2.so: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by lib/jni/x86_64-Linux/libjffi-1.2.so)
	linux-vdso.so.1 =>  (0x00007f413175b000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f413116e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f413153c000)
```

**Comment by headius** [comment]
Looks like this is a [pretty common problem](https://www.google.com/search?q=version+%60GLIBC_2.27%27+not+found&oq=version+%60GLIBC_2.27%27+not+found&aqs=chrome..69i57j0i512l9.222j0j7&sourceid=chrome&ie=UTF-8) for other libraries and runtimes. The library in question for us is the native binary of https://github.com/jnr/jffi, which was probably most recently build and released on a very recent Linux (either Fedora or Ubuntu).

In order to build this we'd need to pin such builds to an old enough glibc to cover major long-term-supported Linux distributions, but also new enough that it would remain compatible on the newest releases.

**Comment by rubyFeedback** [comment]
It would be kind of great if linux distributions in general could adopt multiple versions of software. Granted, for glibc this is not realistic, as the rest of the system depends on a libc implementation, so one may have to install and retain multiple different versions of ALL programs (see how gobolinux or nixos handle multiple installations of different versions), but if this were possible then users could simply switch between glibc versions AND the required programs depending on it. Obviously the "is now abandoned" issue will kick in eventually, so upgrades may have to happen, but it could alleviate some pain if backwards compatibility would be more easily supported by all linux distributions.

I am on 2.36 myself, mostly because I am fine with the latest glibc usually. 2.27 is quite old - https://ftp.gnu.org/gnu/glibc/?C=M;O=D - **2018**.

**Comment by headius** [comment]
So RHEL is just that far behind on updating glibc?

**Comment by UweKubosch** [comment]
We are trying to upgrade our production system, and it is on GLIBC 2.12, Centos 6.  I guess we will have to upgrade the server before upgrading to JRuby 9.4.

**Comment by jsvd** [comment]
We're getting the same issue in Debian 9 and Amazon Linux 2

**Comment by jsvd** [comment]
ffi files changed from 9.3.9.0 to 9.3.10.0 are:
```
lib/jni/aarch64-Linux/libjffi-1.2.so  
lib/jni/arm-Linux/libjffi-1.2.so        
lib/jni/i386-Linux/libjffi-1.2.so       
lib/jni/x86_64-Linux/libjffi-1.2.so  
```

**Comment by jsvd** [comment]
Also seems that these files are packed into the jruby.jar file, so the ones in `lib/jni/` are ignored.

**Comment by headius** [comment]
This will be partially fixed by an update x86_64 jffi build in 9.4.3. Other architectures still need work (see https://github.com/jnr/jffi/issues/138).

**Comment by headius** [comment]
Let's call this one fixed, since it only involved x86_64. The newer issue #7695 will be remain open to fix the other architectures.
